### PR TITLE
Bug 1086930 - Add status region to sync calendar sync button

### PR DESCRIPTION
**Pull request by jobara**
```
Added a status area inside of the sync button this is only added to the a11y tree during syncing. It will specify that syncing has started and completed.
```

### CODE REVIEW

**Comment by try-server-hook** [comment]
[jobara Justin Obara (jobara)](https://github.com/jobara) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=51ff93b63c87)

**Comment by jobara** [comment]
@yzen I've updated based on your review comments. I still need to update the tests but thought I'd push up so you could try it out. It seems that currently you have to move away and comeback to have the progress bar state read by the screen reader. I assume this is because of https://bugzilla.mozilla.org/show_bug.cgi?id=1105770 but thought I'd double check. This of course means that the complete state is never announced.

File: apps/calendar/test/unit/views/settings_test.js [hunk 1]
```diff
@@ -170,6 +172,10 @@ suiteGroup('views/settings', function() {
     assert.ok(subject.syncProgressTarget);
   });
 
+  test('#syncStatus', function() {
```
Comment (yzen) [inline apps/calendar/test/unit/views/settings_test.js]: All syncStatus should be renamed to syncProgress (as in the actual view)

File: apps/calendar/locales/calendar.en-US.properties [hunk 2]
```diff
@@ -79,6 +79,10 @@ remove-account-dialog-details=Removing this account will also remove your local
 # for accessibility purposes, not visible to the eye.
 drawer-sync-button=Sync
 
+# sync progress
```
Comment (yzen) [inline apps/calendar/locales/calendar.en-US.properties]: Add a comment like this: https://github.com/jobara/gaia/blob/1086930/apps/calendar/locales/calendar.en-US.properties#L254-L255 here

File: apps/calendar/test/unit/views/settings_test.js [hunk 3]
```diff
@@ -345,7 +351,13 @@ suiteGroup('views/settings', function() {
     };
```
Comment (yzen) [inline apps/calendar/test/unit/views/settings_test.js]: Also check before triggering the initial state

**Comment by try-server-hook** [comment]
[jobara Justin Obara (jobara)](https://github.com/jobara) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=be08b7051ade)

File: apps/calendar/test/unit/views/settings_test.js [hunk 4]
```diff
@@ -84,6 +84,8 @@ suiteGroup('views/settings', function() {
       '        <button class="sync update toolbar-item">',
       '          <span class="icon"',
       '                data-l10n-id="drawer-sync-button">Sync</span>',
+      '          <span class="sync-progress" role="progressbar"',
+      '                aria-relevant="additions text"></span>',
```
Comment (yzen) [inline apps/calendar/test/unit/views/settings_test.js]: need to remove aria-relevant="additions text"

File: apps/calendar/test/unit/views/settings_test.js [hunk 5]
```diff
@@ -344,8 +350,23 @@ suiteGroup('views/settings', function() {
       calledWith = arguments;
     };
 
+    assert.notOk(subject.syncProgress.getAttribute('data-l10n-id'));
+    assert.ok(!subject.syncProgress.classList.contains('syncing'));
```
Comment (yzen) [inline apps/calendar/test/unit/views/settings_test.js]: nit: notOk ?

File: apps/calendar/test/unit/views/settings_test.js [hunk 6]
```diff
@@ -344,8 +350,23 @@ suiteGroup('views/settings', function() {
...
+    controller.on('syncStart', function () {
+      assert.ok(subject.syncProgress.classList.contains('syncing'));
+      assert.equal(subject.syncProgress.getAttribute('data-l10n-id'),
+        'sync-progress-syncing');
+    });
+
+    controller.on('syncComplete', function () {
+      assert.equal(subject.syncProgress.getAttribute('data-l10n-id'),
+        'sync-progress-complete');
+      assert.ok(!subject.syncProgress.classList.contains('syncing'));
```
Comment (yzen) [inline apps/calendar/test/unit/views/settings_test.js]: nit: notOk ?

**Comment by try-server-hook** [comment]
[jobara Justin Obara (jobara)](https://github.com/jobara) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=c739ade34c37)

**Comment by try-server-hook** [comment]
[jobara Justin Obara (jobara)](https://github.com/jobara) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=edfdfadfdf77)

File: apps/calendar/test/unit/views/settings_test.js [hunk 7]
```diff
@@ -344,8 +349,23 @@ suiteGroup('views/settings', function() {
...
+    assert.notOk(subject.syncProgress.getAttribute('data-l10n-id'));
+    assert.notOk(subject.syncProgress.classList.contains('syncing'));
+
     triggerEvent(subject.syncButton, 'click');
     assert.ok(calledWith);
+
+    controller.on('syncStart', function () {
+      assert.ok(subject.syncProgress.classList.contains('syncing'));
+      assert.equal(subject.syncProgress.getAttribute('data-l10n-id'),
+        'sync-progress-syncing');
```
Comment (millermedeiros) [inline apps/calendar/test/unit/views/settings_test.js]: these assertions are not working the way you expect.. `controller.on('syncStart', fn)` will be triggered asynchronously, so even if the assertions fails they should only fail after the `test` is already complete..

the proper way to test this is to create 2 async tests like:

```js
suite('syncProgress', function() {
  var syncController = app.syncController;

  test('syncStart', function(done) {
    syncController.on('syncStart', function () {
      assert.ok(subject.syncProgress.classList.contains('syncing'));
      assert.equal(subject.syncProgress.getAttribute('data-l10n-id'),
      'sync-progress-syncing');
      done();
    });
    syncController.emit('syncStart');
  });

  test('syncComplete', function(done) {
    syncController.on('syncComplete', function () {
      assert.equal(subject.syncProgress.getAttribute('data-l10n-id'),
      'sync-progress-complete');
      assert.notOk(subject.syncProgress.classList.contains('syncing'));
      done();
    });
    syncController.emit('syncComplete');
  });
});
```

or if you are really want to test the `click` event, you should use the [`waitFor` pattern](https://github.com/jobara/gaia/blob/1086930/apps/calendar/test/unit/controllers/notifications_test.js#L93-L98)

**Comment by try-server-hook** [comment]
[jobara Justin Obara (jobara)](https://github.com/jobara) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=8593db829b2b)

File: apps/calendar/test/unit/views/settings_test.js (hunk 7 above)
Comment (jobara) [inline apps/calendar/test/unit/views/settings_test.js]: @millermedeiros thanks for the explanation. I made the changes you recommended as it's probably enough in this case to test the actions of the events. The only change I made was to call app.syncController directly in the test blocks as app does not seem to be in scope for the suite.

**Pull request merged**